### PR TITLE
Added ACME integration to create new certificates

### PR DESCRIPTION
**Pull request by mattjackson220**
## What does this PR (Pull Request) do?

- [x] This PR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Ops
- Traffic Portal

## What is the best way to verify this PR?

Set up an ACME account with a provider.  
Go to a Delivery Service that has HTTPS enabled, go to manage SSL Keys, Generate SSL Keys
Click on the Use ACME slider
Choose your ACME provider from the drop down
Generate keys and verify that you got a new certificate from your ACME provider
Verify that the Let's Encrypt functionality still works as before
Verify that the /acme_accounts/providers endpoint returns all ACME providers in the cdn.conf file as well as Let's Encrypt

## The following criteria are ALL met by this PR

No tests added since this all depends on 3rd parties
- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_ops/traffic_ops_golang/deliveryservice/acme.go [hunk 1]
```diff
@@ -22,178 +22,408 @@ package deliveryservice
...
+			return errors.New("Zero rows affected when deleting dns txt record for fqdn '" + fqdn + "' record '" + value)
+		}
+	}
+
+	return nil
+}
+
+// GenerateAcmeCertificates gets and saves certificates using ACME protocol from a give ACME provider.
+func GenerateAcmeCertificates(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme.go:151]: This handler should return a 500 if Traffic Vault is not configured

File: traffic_ops/traffic_ops_golang/deliveryservice/acme.go [hunk 2]
```diff
@@ -22,178 +22,408 @@ package deliveryservice
...
+
+// GenerateLetsEncryptCertificates gets and saves new certificates from Let's Encrypt.
+func GenerateLetsEncryptCertificates(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
 	if userErr != nil || sysErr != nil {
-		api.HandleErr(w, r, inf.Tx.Tx, statusCode, userErr, sysErr)
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme.go:210]: This handler should return a 500 if Traffic Vault is not configured

**Review by rawlinp — CHANGES_REQUESTED** [review]
I just did a quick scan since I'm working on refactoring the Riak integration which will have some conflicts with this.

File: traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		return nil, errors.New("getting acme client: " + err.Error()), http.StatusInternalServerError
+	}
+
+	renewRequest := certificate.Resource{
+		Certificate: []byte(keyObj.Certificate.Crt),
+	}
+
+	cert, err := client.Certificate.Renew(renewRequest, true, false)
+	if err != nil || cert == nil {
+		log.Errorf("Error obtaining acme certificate: %s", err.Error())
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go:142]: `err` might be nil here

File: traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		Version:         keyObj.Version + 1,
+	}
+
+	newCertObj.Certificate = tc.DeliveryServiceSSLKeysCertificate{
+		Crt: string(EncodePEMToLegacyPerlRiakFormat(cert.Certificate)),
+		Key: string(EncodePEMToLegacyPerlRiakFormat(cert.PrivateKey)),
+		CSR: string(EncodePEMToLegacyPerlRiakFormat([]byte("ACME Generated"))),
+	}
+
+	if err := riaksvc.PutDeliveryServiceSSLKeysObj(newCertObj, tx, cfg.RiakAuthOptions, cfg.RiakPort); err != nil {
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go:161]: post-rebase, once this function takes a `tv trafficvault.TrafficVault`, you would call `tv.PutDeliveryServiceSSLKeys(newCertObj, tx)`

File: traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		log.Errorf("Error posting acme certificate to riak: %s", err.Error())
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go:162]: s/riak/Traffic Vault/

File: traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: FAILED to add SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+		return nil, errors.New(dsName + ": putting riak keys: " + err.Error()), http.StatusInternalServerError
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/acme_renew.go:164]: s/riak/Traffic Vault/

File: traffic_portal/app/src/common/modules/form/deliveryServiceSslKeys/generate/FormGenerateDeliveryServiceSslKeysController.js [hunk 7]
```diff
@@ -334,7 +352,8 @@ var FormGenerateDeliveryServiceSslKeysController = function(deliveryService, ssl
             }
         });
         modalInstance.result.then(function() {
-            deliveryServiceSslKeysService.generateSslKeysWithLetsEncrypt(deliveryService, sslKeys, sslRequest).then(
+			sslKeys.authType = $scope.acmeProvider;
```
Comment (rawlinp) [inline traffic_portal/app/src/common/modules/form/deliveryServiceSslKeys/generate/FormGenerateDeliveryServiceSslKeysController.js:355]: indentation is off

File: traffic_portal/app/src/common/modules/form/deliveryServiceSslKeys/generate/form.GenerateDeliveryServiceSslKeys.tpl.html [hunk 8]
```diff
@@ -91,9 +91,17 @@
                     <span ng-show="hasError(dsGenerateSslKeyForm.commonName)" class="form-control-feedback"><i class="fa fa-times"></i></span>
                 </div>
             </div>
+			<div class="form-group" ng-class="{'has-error': hasError(dsGenerateSslKeyForm.country), 'has-feedback': hasError(dsGenerateSslKeyForm.country)}" ng-show="useAcme">
```
Comment (rawlinp) [inline traffic_portal/app/src/common/modules/form/deliveryServiceSslKeys/generate/form.GenerateDeliveryServiceSslKeys.tpl.html:94]: indentation is off

File: traffic_portal/app/src/scripts/config.js [hunk 9]
```diff
@@ -23,6 +23,6 @@
 
 angular.module('config', [])
 
-.constant('ENV', { api: { root:'/api/4.0/', legacy: '/api/1.5/' } })
+.constant('ENV', { api: { root:'/api/4.1/', legacy: '/api/1.5/' } })
```
Comment (rawlinp) [inline traffic_portal/app/src/scripts/config.js:26]: I think we're still on 4.0

**Review by rawlinp — CHANGES_REQUESTED** [review]
Sorry for all the comments but they were mostly just the same things repeated. If you need help w/ the rebase due to the TrafficVault changes let me know. Basically wherever `riaksvc` was being used, you now need to pass in a `trafficvault.TrafficVault` instance to call instead.

**Comment by mattjackson220** [comment]
Closing and replacing with https://github.com/apache/trafficcontrol/pull/5728 due to case sensitivity issues

**Pull request closed without merging**